File: internal/instancesvc/events.go
package instancesvc

import (
	"sync"
	"time"
)

// 事件类型。前端通过 SSE 订阅它们来实时更新界面，
// 而不是靠轮询——生命周期转换要 30–90 秒，轮询既慢又浪费。
const (
	EventInstanceUpdated = "instance.updated"
	EventInstanceRemoved = "instance.removed"
	EventInstanceError   = "instance.error"
	EventAccountStatus   = "account.status"
	EventSyncStarted     = "sync.started"
	EventSyncFinished    = "sync.finished"
)

// Event 是一条推给前端的实时事件。
type Event struct {
	Type       string    `json:"type"`
	At         time.Time `json:"at"`
	InstanceID string    `json:"instanceId,omitempty"`
	AccountID  string    `json:"accountId,omitempty"`
	State      string    `json:"state,omitempty"`
	Message    string    `json:"message,omitempty"`
	Data       any       `json:"data,omitempty"`
}

// subBuffer 是每个订阅者的缓冲深度。
//
// 一次全量同步可能瞬间产生几十条事件，缓冲太浅会让正常的浏览器订阅
// 也开始丢事件；太深则会在客户端卡死时占住内存。
const subBuffer = 64

// Bus 是进程内的事件广播。
//
// 刻意不做持久化：SSE 断线重连后前端会重新拉一次全量列表，
// 补发历史事件既复杂又没有必要。
type Bus struct {
	mu   sync.Mutex
	subs map[int]chan Event
	next int
}

// NewBus 创建事件总线。
func NewBus() *Bus {
	return &Bus{subs: make(map[int]chan Event)}
}

// Subscribe 注册一个订阅者，返回事件通道与取消函数。
// 调用方必须在结束时调用取消函数，否则订阅会一直占着。
func (b *Bus) Subscribe() (<-chan Event, func()) {
	b.mu.Lock()
	defer b.mu.Unlock()

	id := b.next
	b.next++
	ch := make(chan Event, subBuffer)
	b.subs[id] = ch

	return ch, func() {
		b.mu.Lock()
		defer b.mu.Unlock()
		if c, ok := b.subs[id]; ok {
			delete(b.subs, id)
			close(c)
		}
	}
}

// Publish 向所有订阅者广播。
//
// 对满了的订阅者直接丢弃事件而不是阻塞：一个卡住的浏览器标签页
// 绝不能拖慢后台的同步循环。丢了也不要紧——前端本来就会定期兜底刷新。
func (b *Bus) Publish(e Event) {
	if e.At.IsZero() {
		e.At = time.Now()
	}
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, ch := range b.subs {
		select {
		case ch <- e:
		default:
		}
	}
}

// SubscriberCount 返回当前订阅者数量，用于诊断。
func (b *Bus) SubscriberCount() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.subs)
}

File: web/src/api/events.ts
/**
 * SSE 实时事件流。
 *
 * 生命周期转换要 30–90 秒，轮询既慢又浪费。后端用 SSE 单向推送，
 * 浏览器的 EventSource 自带断线重连——但重连后必须重新拉一次全量列表，
 * 因为后端不做历史事件补发（见 docs/API.md）。
 */

export type ServerEventType =
  | 'instance.updated'
  | 'instance.removed'
  | 'instance.error'
  | 'account.status'
  | 'sync.started'
  | 'sync.finished'

export interface ServerEvent {
  type: ServerEventType
  at: string
  instanceId?: string
  accountId?: string
  state?: string
  message?: string
  data?: unknown
}

export interface EventHandlers {
  onEvent?: (event: ServerEvent) => void
  /** 连接建立时触发。断线重连也会触发，此时应重新拉全量数据。 */
  onOpen?: (reconnected: boolean) => void
  onError?: () => void
}

const ALL_TYPES: ServerEventType[] = [
  'instance.updated', 'instance.removed', 'instance.error',
  'account.status', 'sync.started', 'sync.finished'
]

/**
 * 建立事件流连接，返回断开函数。
 *
 * 只在整个应用生命周期内连一条：SSE 在 HTTP/1.1 下占用一个连接，
 * 每个页面各连一条很容易撞上浏览器的并发连接上限。
 */
export function connectEvents(handlers: EventHandlers): () => void {
  let source: EventSource | null = null
  let opened = false
  let closed = false

  const open = () => {
    if (closed) return
    source = new EventSource('/api/events', { withCredentials: true })

    source.onopen = () => {
      const reconnected = opened
      opened = true
      handlers.onOpen?.(reconnected)
    }

    source.onerror = () => {
      // EventSource 会自己重连，这里只用于通知 UI 显示"连接中断"。
      handlers.onError?.()
    }

    for (const type of ALL_TYPES) {
      source.addEventListener(type, (raw) => {
        try {
          const event = JSON.parse((raw as MessageEvent).data) as ServerEvent
          handlers.onEvent?.(event)
        } catch {
          // 事件体解析失败没有补救办法，丢掉即可——
          // 前端本来就有定期兜底刷新。
        }
      })
    }
  }

  open()

  return () => {
    closed = true
    source?.close()
    source = null
  }
}
